### PR TITLE
Circleci project setup

### PR DESCRIPTION
**Pull request by zackbakerusc**
This version includes previous overlay work but passes flake8 and pytest

### CODE REVIEW

**Comment by zackbakerusc** [comment]
Victoriously passing all flake8, pytest, and coveralls testing..

File: plotille/_canvas.py [hunk 1]
```diff
@@ -147,9 +157,18 @@ def _set(self, x_idx, y_idx, set_=True, color=None):
         y_c, y_p = y_idx // 4, y_idx % 4
 
         if 0 <= x_c < self.width and 0 <= y_c < self.height:
-            self._canvas[y_c][x_c].update(x_p, y_p, set_)
-            if color:
-                self._canvas[y_c][x_c].fg = color
+            if(not overlay):
```
Comment (tammoippen) [inline plotille/_canvas.py:160]: No parentheses here and in all other if places, if not necessary.

File: plotille/_canvas.py [hunk 2]
```diff
@@ -248,8 +268,20 @@ def plot(self, linesep=linesep):
             linesep: str  The requested line seperator. default: os.linesep
 
         Returns:
-            unicode: The canvas as a string.
+            unicode: The canvas as a string, overlaid with textual elements
         """
 
-        return linesep.join(''.join(map(six.text_type, row))
-                            for row in reversed(self._canvas))
+        # classic:
```
Comment (tammoippen) [inline plotille/_canvas.py:274]: Remove legacy comments.

File: plotille/_canvas.py [hunk 3]
```diff
@@ -248,8 +268,20 @@ def plot(self, linesep=linesep):
...
+        # ret = linesep.join(''.join(map(six.text_type, row)) for row in reversed(self._canvas))
+        # return ret
+
+        # combine _canvas (Dots) and overlay (text) so that overlay is on top of dots
+        ret_combined = ''
+        for row_index in range(len(self._canvas))[::-1]:
+            for ii in range(len(self._canvas[row_index])):
+                if(self._canvas_overlay[row_index][ii] == ' '):
```
Comment (tammoippen) [inline plotille/_canvas.py:282]: This means, if the text in the overlay has a `' '`, we might have dots between the text?

File: plotille/_figure.py [hunk 4]
```diff
@@ -28,6 +28,7 @@
 from itertools import cycle
 import os
 
+import numpy as np
```
Comment (tammoippen) [inline plotille/_figure.py:31]: This is not a dependency of `plotille` and i would like to keep it that way. Please implement with stdlib funktions.

File: plotille/_figure.py [hunk 5]
```diff
@@ -238,17 +239,56 @@ def _x_axis(self, xmin, xmax, label='X', with_y_axis=False):
...
     def plot(self, X, Y, lc=None, interp='linear', label=None):  # noqa: N803
         if len(X) > 0:
             if lc is None:
                 lc = next(self._color_seq)[self.color_mode]
-            self._plots += [Plot.create(X, Y, lc, interp, label)]
+            overlay = False
+            self._plots += [Plot.create(X, Y, lc, interp, label, overlay)]
 
-    def scatter(self, X, Y, lc=None, label=None):  # noqa: N803
+    def scatter(self, X, Y, lc=None, label=None, marker='', text=None):  # noqa: N803
```
Comment (tammoippen) [inline plotille/_figure.py:281]: Better use `None` to indicate the field is not used?

```suggestion
    def scatter(self, X, Y, lc=None, label=None, marker=None, text=None):  # noqa: N803
```

The test later can be something like: `if marker:`

File: plotille/_figure.py [hunk 6]
```diff
@@ -324,24 +363,39 @@ def height_vals(self):
         return self.Y
 
     def write(self, canvas, with_colors, in_fmt):
+
+        # plot all points with optional text first, then lines
+        # separating out points from lines allows single-point scatters to successfuly plot
+
+        # make point iterators
```
Comment (tammoippen) [inline plotille/_figure.py:370]: Comment is wrong / old.

File: requirements.txt [hunk 7]
```diff
@@ -0,0 +1,5 @@
+numpy==1.15.3
```
Comment (tammoippen) [inline requirements.txt:1]: Remove file. Dependency management is done with [poetry](https://python-poetry.org/)

File: plotille/_figure.py [hunk 8]
```diff
@@ -324,24 +363,39 @@ def height_vals(self):
...
+        points = zip(map(in_fmt.convert, self.X), map(in_fmt.convert, self.Y))
+
+        for index, (x, y) in enumerate(points):
+
+            if(isinstance(self.text, list) and index <= len(self.text)):
+                text_for_point = self.text[index]
+            else:
+                text_for_point = self.text
+
+            canvas.point(x, y, color=color, overlay=self.overlay, marker=self.marker, text=text_for_point)
```
Comment (tammoippen) [inline plotille/_figure.py:382]: The `text` of `Plot` is used for every point? This should be the `label` for the legend, right?

File: plotille/_figure.py [hunk 9]
```diff
@@ -324,24 +363,39 @@ def height_vals(self):
...
         next(to_points)
 
-        color = self.lc if with_colors else None
         # plot points
         for (x0, y0), (x, y) in zip(from_points, to_points):
-            canvas.point(x0, y0, color=color)
-
-            canvas.point(x, y, color=color)
             if self.interp == 'linear':
                 canvas.line(x0, y0, x, y, color=color)
```
Comment (tammoippen) [inline plotille/_figure.py:394]: Can it be, that the line interpolations overwrite the text from above?

Also, if the plotting of points is moved to the code above, I would put all the code from 384ff into the `if self.interp == 'linear':` block.

**Comment by zackbakerusc** [comment]
None of that should be a problem.  I'll get on it this weekend.

**Comment by tammoippen** [comment]
Closing as #38 implements most of the features.

**Pull request closed without merging**